src/gin/nccl_ofi_gin_reqs.cpp

```diff
@@ -0,0 +1,270 @@
+/*
+ * Copyright (c) 2025      Amazon.com, Inc. or its affiliates. All rights reserved.
+ */
+
+#include "config.h"
+
+#include "gin/nccl_ofi_gin.h"
+#include "gin/nccl_ofi_gin_reqs.h"
+#include "gin/nccl_ofi_gin_resources.h"
+
+int nccl_net_ofi_gin_op_req_t::ctx_handle_cq_entry
+		(nccl_net_ofi_gin_context *ctx,
+		 struct fi_cq_entry *cq_entry_base,
+		 fi_addr_t src_addr,
+		 uint16_t rail_id)
+{
+	nccl_net_ofi_gin_op_req_t *req = cpp_container_of(ctx, &nccl_net_ofi_gin_op_req_t::ctx);
+	return req->handle_cq_entry(cq_entry_base, src_addr, rail_id);
+}
+
+int nccl_net_ofi_gin_op_req_t::ctx_handle_error_entry
+		(nccl_net_ofi_gin_context *ctx,
+		 struct fid_cq *cq,
+		 struct fi_cq_err_entry *err_entry,
+		 uint16_t rail_id)
+{
+	int ret = 0;
+
+	if (err_entry->err == FI_ECANCELED) {
+		/* Closing an EP with posted receives will (erroneously) generate
+		   cancellation events for the posted receives with the EFA provider
+		   in Libfabric versions prior to 1.22. These events are harmless
+		   and can be ignored.
+
+		   With Libfabric 1.22 and later, we shouldn't get these cancel
+		   events at all. The plugin does not explicitly call fi_cancel. */
+		ret = -(err_entry->err);
+		return ret;
+	}
+
+	assert(ctx);
+	nccl_net_ofi_gin_op_req_t *req = cpp_container_of(ctx, &nccl_net_ofi_gin_op_req_t::ctx);
+
+	NCCL_OFI_WARN("Request %p completed with error. RC: %d. Error: %d (%s). Completed length: %ld",
+		req, err_entry->err,
+		err_entry->prov_errno,
+		fi_cq_strerror(cq, err_entry->prov_errno, err_entry->err_data, NULL, 0),
+		(long)err_entry->len);
+
+	/*
+	 * Libfabric error codes directly map to ISO C errno values for standard
+	 * error codes up to FI_ERRNO_OFFSET, and libfabric-specific error codes
+	 * beyond. nccl_net_ofi_retval_translate() will figure out
+	 * how to deal with these, so it is safe to pass up the err as-is.
+	 * However, any special-handling for prov_errno should be handled here.
+	 */
+	ret = -(err_entry->err);
+	return ret;
+}
+
+nccl_net_ofi_gin_op_req_t::nccl_net_ofi_gin_op_req_t() :
+	ctx({.ofi_ctx = {},
+	     .handle_cq_entry = ctx_handle_cq_entry,
+	     .handle_error_entry = ctx_handle_error_entry})
+{ }
+
+
+/* Receive requests */
+nccl_net_ofi_gin_recv_req_t::nccl_net_ofi_gin_recv_req_t(nccl_ofi_gin_resources &resources_arg,
+							 nccl_ofi_gin_ep_rail_t &rail_arg)
+	: nccl_net_ofi_gin_op_req_t(),
+	  resources(resources_arg),
+	  rail(rail_arg)
+{
+	rx_buff_elem = nccl_ofi_freelist_entry_alloc(resources.get_rx_buff_fl());
+	if (!rx_buff_elem) {
+		NCCL_OFI_WARN("Failed to allocate rx buffer freelist entry");
+		throw std::runtime_error("Failed to allocate rx buffer freelist entry");
+	}
+}
+
+
+nccl_net_ofi_gin_recv_req_t::~nccl_net_ofi_gin_recv_req_t()
+{
+	nccl_ofi_freelist_entry_free(resources.get_rx_buff_fl(), rx_buff_elem);
+}
+
+
+int nccl_net_ofi_gin_recv_req_t::handle_cq_entry(struct fi_cq_entry *cq_entry_base,
```

Comment:
A general question: I see `int` being using in most of the places. Is there a hard requirement by NCCL/EFA that `int` needs to be used instead of a datatype that is explicit in its sign-ness and size such as like `uint32_t`?

Reply:
`int` is always a signed type in C++.  There is no ambiguity there.  This is going to return an errno-style error code, which are all integers in Unix convention.  So `int` makes the most sense to me.